src/EditorFeatures/Core/Workspaces/ITextBufferVisibilityTracker.cs

```diff
@@ -37,40 +39,59 @@ internal interface ITextBufferVisibilityTracker
-            if (service.IsVisible(subjectBuffer))
-                return;
+            // Because cancellation is both expensive, and a super common thing to occur while we're delaying the caller
+            // until visibility, we special case the implementation here and transition to the canceled state
+            // explicitly, rather than throwing a cancellation exception.
 
-            // ensure we listen for visibility changes before checking.  That way we don't have a race where we check
-            // something see it is not visible, but then do not hear about its visibility change because we've hooked up
-            // our event after that happens.
-            var visibilityChangedTaskSource = new TaskCompletionSource<bool>();
-            var callback = void () => visibilityChangedTaskSource.TrySetResult(true);
-            service.RegisterForVisibilityChanges(subjectBuffer, callback);
+            var taskOfTask = DelayWhileNonVisibleWorkerAsync().ContinueWith(
+                task => cancellationToken.IsCancellationRequested ? Task.FromCanceled(cancellationToken) : task,
+                CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
+            return taskOfTask.Unwrap();
 
-            try
+            // Normal delay logic, except that this does not throw in the event of cancellation, but instead returns
+            // gracefully.  The above task continuation logic then ensures we return a canceled task without needing
+            // exceptions.
+            async Task DelayWhileNonVisibleWorkerAsync()
             {
-                // Listen to when the active document changed so that we startup work on a document once it becomes visible.
-                var delayTask = Task.Delay(timeSpan, cancellationToken);
-                await Task.WhenAny(delayTask, visibilityChangedTaskSource.Task).ConfigureAwait(true);
-            }
-            finally
-            {
-                service.UnregisterForVisibilityChanges(subjectBuffer, callback);
+                if (cancellationToken.IsCancellationRequested)
+                    return;
+
+                await threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                if (service.IsVisible(subjectBuffer))
+                    return;
+
+                // ensure we listen for visibility changes before checking.  That way we don't have a race where we check
+                // something see it is not visible, but then do not hear about its visibility change because we've hooked up
+                // our event after that happens.
+                var visibilityChangedTaskSource = new TaskCompletionSource<bool>();
+                var callback = void () => visibilityChangedTaskSource.TrySetResult(true);
+                service.RegisterForVisibilityChanges(subjectBuffer, callback);
+
+                try
```

Comment:
I assume utilizing CancellationToken.Register and doing something like:

cancellationToken.Register(callback);

and then sending in CancellationToken.None to the IAsynchronousOperationListener.Delay call isn't an option to prevent the exception from being thrown when Task.Delay is inside the listener.Delay call?

Reply:
This is a very nice idea.  I wasn't able to apply it though here as we still need the CT for Expeditable.Delay.  Maybe sam has thoughts on how to combine both of htese?